### PR TITLE
monit - run monit scripts inside docker containers

### PR DESCRIPTION
**Pull request by mapellidario**
Related to #6493 

### status

Tested, ready to be reviewed

### Description

The purpose of this PR is to run the scripts GenerateMonit, CheckTapeRecall, aso_metrics_ora and ReportRecallQuota inside docker containers.

- GenerateMonit.py runs inside a crabtaskworker container.
- CheckTapeRecall.py, aso_metrics_ora and ReportRecallQuota runs inside a python:3.8 

design choices:

- run script inside a docker container. The script runs once per `docker run`
- script output:
    - every script has it's own application log, same as before
    - every script prints to stdout. if the script fails, then the stdout is saved in a file in `/tmp/monit-*.txt`
    - [ ] ? shall we get rid of the application log and send everything to stdout? 

next steps

- [ ] merge this PR
- [ ] change puppet qa and master so that runcontainer.sh is taken from master branch (not from the branch of this PR)
- [ ] disable scripts on crab-preprod-tw02, enable script on crab-prod-tw02
- [ ] disable recurring action [report recall quota](https://github.com/dmwm/CRABServer/blob/master/src/python/TaskWorker/Actions/Recurring/ReportRecallQuota.py)

All further information is available in our docs, at:

- https://cmscrab.docs.cern.ch/technical/crab-monitoring/crab-opensearch.html
- https://cmscrab.docs.cern.ch/technical/crab-monitoring/crab-crontabs.html

### CODE REVIEW

**Comment by belforte** [comment]
Very nice comments Dario. But I fear that they will get lost. Maybe this is something for https://cmscrab.docs.cern.ch/ ? Indeed the Monitoring section in there needs expanding and we lack a "How to develop" which is where advice on e.g. containers should better fit

**Comment by mapellidario** [comment]
GenerateMonit is properly running from the crontab inside a docker container [0] on `crab-dev-tw04` and the result looks good [1]

If nobody disagress, on monday I will merge this PR. 

I will make the required changes to run these scripts in "production" on another PR, which will need to be synchronized with changes in puppet.

---

[0] 

```plaintext
# Puppet Name: Send information to elasticsearch MONIT
*/5 * * * * test -f /home/crab3/monit_run-taskworker.sh && bash /home/crab3/monit_run-taskworker.sh >/dev/null 2>&1
```

[1] https://monit-opensearch.cern.ch/dashboards/goto/4e055aa73b4e128250e001d032243334

**Comment by belforte** [comment]
I do not understand why you need a new script instead of creating a new service in runContainer.sh
IIUC the "specialization of the container" is via the tag ( -v option to runContainer.sh)

File: src/script/Container/runContainer.sh [hunk 1]
```diff
@@ -84,6 +84,7 @@ DOCKER_VOL="${DOCKER_VOL} -v /etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem:/
 DOCKER_OPT="-e SERVICE=${SERVICE} -w /data/srv/${DIRECTORY} "
 
 if [[ "${SERVICE}" == TaskWorker_monit_*  ]]; then
+  echo "TaskWorkerr_monit_* detected"
```
Comment (belforte) [inline src/script/Container/runContainer.sh:87]: typo: Workerr --> Worker (one only 'r')

**Comment by mapellidario** [comment]
This PR is ready to be reviewed, all additional information is in our mkdocs. Thanks to everybody for the patience!

File: cicd/build/monit-python/monit_build2.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
```
Comment (mapellidario) [inline cicd/build/monit-python/monit_build2.sh:2]: @novicecpp do we prefer this file or `cicd/build/monit-python/monit_build.sh` ? :) 
(sorry, I forgot that i could copy how you write dockerfiles for pypi :) )

**Comment by mapellidario** [comment]
I just noticed that this PR supersedes the recurring action defined in https://github.com/dmwm/CRABServer/blob/master/src/python/TaskWorker/Actions/Recurring/ReportRecallQuota.py

**Review by belforte — APPROVED** [review]

**Review by novicecpp — APPROVED** [review]
Thanks Dario. LGTM.

Quick question: could you please remind me why we have 2 containers?

File: cicd/build/monit-python/monit_build2.sh (hunk 2 above)
Comment (novicecpp) [reply cicd/build/monit-python/monit_build2.sh:2]: Sorry, I am not understand the question. Maybe it is not relevant to the latest change anymore?

**Comment by belforte** [comment]
> Quick question: could you please remind me why we have 2 containers?

:open_mouth:  good question, I had not noticed. Hmm.. maybe I did not read carefully. Sorry, so many things.

File: cicd/build/monit-python/monit_build2.sh (hunk 2 above)
Comment (mapellidario) [reply cicd/build/monit-python/monit_build2.sh:2]: I wrote `monit_build.sh` before i noticed that I did not need to copy files in a subdirectory. I could just build the docker container from the dmwm/CRABServer root and use that directory as build as context. I will clean this up, assuming that you prefer `monit_build2.sh` more than `monit_build.sh`, as I do now, some weeks after creating it  :)

Comment (belforte) [reply cicd/build/monit-python/monit_build2.sh:2]: so that directory will eventually have only one `Dockerfile` and only one `monit_build.sh` file ?

**Comment by belforte** [comment]
> Since I did not manage to install numpy on the TW container using the spec files, I decided that the other scripts can be run from a plain usual debian python docker container

Ah.. I remember now. Sorry that I had forgotten.

**Comment by novicecpp** [comment]
Ah, numpy and the RPM....
Thanks. I remember we already had this discussion in Weekly meeting but cannot recall why.

**Comment by mapellidario** [comment]
I cleaned up a bit the dockerfiles and the build scripts. now it is ready to merge. thanks to everybody for the review and for the patience! :)

**Pull request merged**